extensions/panache/mongodb-panache-common/runtime/src/main/java/io/quarkus/mongodb/panache/common/reactive/runtime/ReactiveMongoOperations.java

```diff
@@ -287,10 +312,15 @@ private Uni<Void> persistOrUpdate(ReactiveMongoCollection collection, List<Objec
                 //insert with autogenerated ID
                 bulk.add(new InsertOneModel(entity));
             } else {
-                //insert with user provided ID or update
-                BsonDocument query = new BsonDocument().append(ID, id);
-                bulk.add(new ReplaceOneModel(query, entity,
-                        new ReplaceOptions().upsert(true)));
+                //insert/update with user provided ID or update
+                VersionHandler versionHandler = VersionHandler.handle(entity, document);
+                if (versionHandler.containsVersionAnnotationAndValue()) {
+                    //it will ignore in silence not matching updates.
+                    bulk.add(new ReplaceOneModel(versionHandler.query, entity));
+                } else if (!versionHandler.containsVersionValue) {
+                    //only try upsert when there is no version value
```

Comment:
I've changed to handle this scenario, when we have versioned entities, we call the `update` method that check version and throw the OptimisticLockException, otherwise I call `collection.bulkWrite(bulk);`.
The drawback here is to no been able to call `bulkWrite` for all and get noticed of every single entity that could be failed in the update.
Do you have any suggestion to better handle this?

Reply:
I've changed this, if we have version annotation, it follows another path and the state is guaranteed